### DISCUSSION TITLE
No Build & Image with Pull Policy

### DISCUSSION THREAD

**Discussion by adaliszk** [zavoloklom/docker-compose-linter#38] [General]
While the rule makes a lot of sense to limit the unintended behaviours even if it has a fairly stable logic, it also acknowledges that with a `pull_policy` the ambiguity can be controlled.

> If pull_policy is missing from the service definition, Compose attempts to pull the image first and then builds from source if the image isn't found in the registry or platform cache.

It would be nice to allow the build and image keys together if we also declare the pull policy to address the concerns of the rule.

The reason for the desire to have both keys is simply that with large teams and multiple/mono -repos, it is very useful to not expect that every single build step would be compiled on every developers machine. In my experience, using the image while providing a build creates time-savings in:

- Initial repository setup; not often needed but when you may spend minutes to clone repositories, it does encourages people to see entire repos to boot up extremely fast after the initial wait time.
- Code changes with smart cache layers; when eventually developers do run into the need to build, or processed force `--build` flags, it is really nice to skip 80% of a Dockerfile because everyone already has the cache layers up until the important COPY lines.

**Comment by zavoloklom** [comment]
Hi @adaliszk!
I assume that in this case it's easier to turn off this rule in configuration. Why do you think it's better to check existence `pull_policy` instead?

**Reply by adaliszk** [reply]
While I can disable this for my projects, I wanted to open a discussion about the rule itself as it suggests on its own that there is a way where having both the build and the image is okay, but the tool does not check for that condition.

It feels like a missed feature or a wrongly explained reasoning.

**Reply by zavoloklom** [reply]
I see, what do you think about making this an option for this rule? Something like "checkPullPolicy": true/false ?

**Reply by adaliszk** [reply]
I like Opt-in paths much better than Opt-out ones. Especially if it can be monitored and how it is used in the wild.

**Comment by zavoloklom** [comment]
@adaliszk Hi! I add this feature to new version - https://github.com/zavoloklom/docker-compose-linter/releases/tag/v2.0.0

Documentation - https://github.com/zavoloklom/docker-compose-linter/blob/main/docs/rules/no-build-and-image-rule.md#options
